File: 02/nim/part01.nim
import strutils, sequtils

let file = open("./inputs/02.txt")
var input = readAll(file).split({ ',' }).mapIt(parseInt(strip(it)))
file.close()

input[1] = 12
input[2] = 2

var
  pos     = 0
  halted  = false
  opcode  = input[0]

while not halted:
  if opcode == 1:
    input[input[pos + 3]] = input[input[pos + 1]] + input[input[pos + 2]]
    pos += 4

  if opcode == 2:
    input[input[pos + 3]] = input[input[pos + 1]] * input[input[pos + 2]]
    pos += 4

  if opcode == 99:
    halted = true

  opcode = input[pos]

echo input[0]

File: 02/ruby/part01.rb
input = File.read('./inputs/02.txt').split(',').map(&:to_i)

pos = 0

input[1] = 12
input[2] = 2

loop do
  opcode = input[pos]
  if opcode == 1
    a, b = input[input[pos + 1]], input[input[pos + 2]]
    input[input[pos + 3]] = a + b
    pos += 4
    next
  elsif opcode == 2
    a, b = input[input[pos + 1]], input[input[pos + 2]]
    input[input[pos + 3]] = a * b
    pos += 4
  elsif opcode == 99
    break
  end
end

puts input[0]
